### PR TITLE
Use special strings for methodSig to handle contract call by pre-compiled ABI data

### PR DESCRIPTION
**Pull request by songge-cb**
### Motivation
This PR is to add some logic to the preprocess API to handle fallback pattern contract call. This type of contract call does not have a method signature.

### Solution
We will use either empty string or a unique string "NO-METHOD-SIG" to indicate it. The pre-compiled ABI data will be passed in as methodArgs, either in the string format or the first element of an array.

### Testing

- Tested Preprocess API locally by passing empty string and "NO-METHOD-SIG" as methodSig.

### CODE REVIEW

File: services/construction/contract_call_data.go [hunk 1]
```diff
@@ -80,6 +89,33 @@ func ConstructContractCallDataGeneric(methodSig string, methodArgs interface{})
...
+	case []interface{}:
+		if len(args) == 1 {
+			argStr, isStrVal := args[0].(string)
+			if !isStrVal {
+				return nil, fmt.Errorf("failed to convert method arg \"%T\" to string", args[0])
+			}
+			return argStr, nil
+		}
+		return methodArgs, nil
+	case []string:
```
Comment (xiaying-peng) [inline services/construction/contract_call_data.go:109]: do we need to handle this case?

Comment (songge-cb) [reply services/construction/contract_call_data.go:109]: For this specific project, no. Doing it for generalization.

File: services/construction/contract_call_data.go [hunk 2]
```diff
@@ -80,6 +89,33 @@ func ConstructContractCallDataGeneric(methodSig string, methodArgs interface{})
...
 
+// preprocessArgs converts methodArgs to a string value if methodSig is an empty string.
+// We are calling a contract written with fallback pattern, which has no method signature.
+func preprocessArgs(methodSig string, methodArgs interface{}) (interface{}, error) {
+	if methodSig != "" && methodSig != NoMethodSig {
+		return methodArgs, nil
+	}
+
+	switch args := methodArgs.(type) {
+	case []interface{}:
```
[the remark this answers is not in the document]
Comment (songge-cb) [reply services/construction/contract_call_data.go:100]: As far as I know, no. The arg is the pre-compiled data, which is one piece.

File: services/construction/contract_call_data.go [hunk 3]
```diff
@@ -80,6 +89,33 @@ func ConstructContractCallDataGeneric(methodSig string, methodArgs interface{})
 	}
 }
 
+// preprocessArgs converts methodArgs to a string value if methodSig is an empty string.
+// We are calling a contract written with fallback pattern, which has no method signature.
+func preprocessArgs(methodSig string, methodArgs interface{}) (interface{}, error) {
+	if methodSig != "" && methodSig != NoMethodSig {
```
Comment (xiaying-peng) [inline services/construction/contract_call_data.go:95]: Let's simplify your impl like this:
```
func preprocessArgs(methodSig string, methodArgs interface{}) (interface{}, error) {
	if methodSig == "" || methodSig == NoMethodSig {
		switch args := methodArgs.(type) {
		case []interface{}:
			if len(args) == 1 {
				if argStr, ok := args[0].(string); ok {
					return argStr, nil
				}
				return nil, fmt.Errorf("failed to convert method arg \"%T\" to string", args[0])
			}
		case []string:
			if len(args) == 1 {
				return args[0], nil
			}
		}
	}

	return methodArgs, nil
}
```

Comment (songge-cb) [reply services/construction/contract_call_data.go]: Updated the PR.

**Review by xiaying-peng — APPROVED** [review]

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

**Pull request merged**